### DISCUSSION TITLE
Ephemeral nodes and when and how to obsolete them

### DISCUSSION THREAD

**Discussion by shyamvalsan** [netdata/netdata#13180] [Ideas]
This is my understanding of what an ephemeral node is and how the system currently behaves in terms of node obsoletion - please correct me if/where I am mistaken.

### Definition of Ephemerality
 - Ephemeral nodes (or rather node instances) are created for a specific purpose and the user expects them to not exist after their purpose has been served. Ephemeral nodes that are offline (after their purpose has been served) are not expected to come back online and the user does not want to see or be notified about them.

### Node obsoletion

 **No user action**
   - When a node becomes unreachable or offline it continues to be visible in Netdata cloud
   - If a node has been unreachable for 60 consecutive days it gets (hard) deleted
     - [Planned improvement](https://github.com/netdata/cloud-backend/issues/203): Nodes get automatically soft deleted after 60 consecutive days of being offline, 30 days later they get hard deleted
   
 **User manually obsoletes a node from Netdata cloud**
   - This action is only available for "offline" nodes
   - User goes to nodes tab and selects offline nodes they want to obsolete
   - Obsoleted nodes are soft deleted, which removes the nodes and all charts and alerts associated to it from the UI but leaves the node intact in case it comes back online later.
   - After 60 consecutive days of being offline the node gets (hard) deleted as well.
     - [Planned improvements](https://github.com/netdata/cloud-backend/issues/203): Node gets hard deleted 30 days after user has manually soft deleted it
     - [Planned improvements](https://github.com/netdata/netdata-cloud/issues/464): Users should be able to obsolete offline nodes in bulk through multi select settings
     - [Planned improvements](https://github.com/netdata/netdata-cloud/issues/464): Users should have option to not be warned with popup each time they obsolete a node
     - [Planned improvements](https://github.com/netdata/cloud-frontend/issues/3377): Node obsoletion should be available from manage space/room settings
   
  **NEW/TBD - User marks a node as ephemeral during installation**
   - When an ephemeral node becomes offline it is automatically soft deleted after X minutes (X = ?) without the need for further user intervention
   - Soft deletion removes the nodes and all charts and alerts associated to it from UI but the state is maintained in case it comes back online later.
   - After 60 consecutive days of being offline the node gets (hard) deleted as well.
   

### **Open Questions/Proposals**
 - Will a user (always) know at install-time (or claim-time) if a node is ephemeral? What if they don't?
 - What to do with an ephemeral node if it becomes unreachable due to connectivity issues or reasons other than the node itself finishing its purpose.
 - Visually how should ephemeral nodes be differentiated from other nodes - if at all?
 - What are the implications for Alerts, ML etc if certain nodes are ephemeral? 

- Should we have a recycle bin for obsoleted nodes? 
  - All soft deleted nodes will reside in recycle bin until they are hard deleted
  - If nodes become live they automatically come back from recycle bin
  - Users have option to restore nodes from recycle bin 

- Should nodes that are about to be soft deleted automatically have a visual warning next to them in nodes list (and wherever nodes are listed) that they will be moved to recycle bin in X days?
  - And link them to the documentation as @cakrit suggests [here ](https://github.com/netdata/cloud-backend/issues/203#issuecomment-1158839161)

cc: @netdata/product @ktsaou @cakrit @amalkov @ralphm @ravilushqa @underhood @vkuznecovas @vinnygats @andrewm4894

**Comment by sashwathn** [comment]
Some inputs based on my understanding but it will be good to have a brainstorming on this topic.

> * Will a user (always) know at install-time (or claim-time) if a node is ephemeral? What if they don't?

The user should always know during installation time if a node is ephemeral or not. I think the most common use case is in hyperscale / autoscale environments where nodes are scaled-in / out based on need and there is no real reason for these nodes to show in the nodes list or to display charts / alerts associated with these.

> * What to do with an ephemeral node if it becomes unreachable due to connectivity issues or reasons other than the node itself finishing its purpose.

Usually in Hyperscale environments, the scaling involves a scale-in and scale-out operations.. If a node is intentionally scaled-in, this is meant to be unreachable and should be an Alert and then this node should not be monitored anymore..
We should check if Netdata agents can identify this action..
But if a node goes offline for any other reason (connectivity etc), it should still be alive in our system.

> * Visually how should ephemeral nodes be differentiated from other nodes - if at all?

During their existence, these nodes are like any other but need a different treatment on their obsoletion.

> * What are the implications for Alerts, ML etc if certain nodes are ephemeral?

Again, during their existence, these nodes are like any other but need a different treatment on their obsoletion.

**Comment by hugovalente-pm** [comment]
> Will a user (always) know at install-time (or claim-time) if a node is ephemeral? What if they don't?

I agree with @sashwathn, the user/orchestrator tool should know if a node is ephemeral, so it should be part of the launch template or something that a given flag is associated with the node. 
If, for some reason, a node wasn't marked as ephemeral we could allow and admin user to do it from the UI.

> in hyperscale / autoscale environments where nodes are scaled-in / out based on need and there is no real reason for these nodes to show in the nodes list or to display charts / alerts associated with these.

Not sure in what case you meant this @sashwathn , to not show the node on the list and/or charts and alerts, since on the next points you seem to suggest for these to be seen as a normal node (if they are meant to be online).

> What to do with an ephemeral node if it becomes unreachable due to connectivity issues or reasons other than the node itself finishing its purpose.

If for some reason a node becomes unhealthy won't it be the orchestrator's job to "clean it up" and ensure a new one is spun off to compensate for this?
We should be able to have this information logged so it could be displayed on a feed or something but probably an alert should be triggered if during "X amount of time" we see a constant unhealthy environment - new ephemeral nodes becoming active and others dying constantly

> Visually how should ephemeral nodes be differentiated from other nodes - if at all?
> What are the implications for Alerts, ML etc if certain nodes are ephemeral?

Besides having the flag, adding some breakdown of nodes counter on the home tab and exposing this flag as filtering criteria I don't see other reasons to differentiate.
For ML the only "issue" I see is about the training since the past training models would go away with the node and they might be applicable to a future node coming live but not really sure if/how this could be tackled.

> Should we have a recycle bin for obsoleted nodes?

All seem valid and good suggestions

> Should nodes that are about to be soft deleted automatically have a visual warning next to them in nodes list (and wherever nodes are listed) that they will be moved to recycle bin in X days?

Not sure about the warning on the nodes to get soft deleted, would we expect the user to take some action? 
For the ones that are to be removed from the recycle bin yes, I think this is a good indicator since here users can restore a node from here.

**Reply by sashwathn** [reply]
> Not sure in what case you meant this @sashwathn , to not show the node on the list and/or charts and alerts, since on the next points you seem to suggest for these to be seen as a normal node (if they are meant to be online).
> 
> >

@hugovalente-pm : I was talking about the case once the node is scaled-in. In this case the node will no longer be available and the next time a node is scaled-out it could be a completely different instance.

**Comment by amalkov** [comment]
@shyamvalsan @hugovalente-pm thank you for your input! 
I have a few suggestions/answers:

1. Let's call "Obsoletion" - "SoftDeletion". This is what it is. As a result, we will have less confusion. 
2. HardDelete only SoftDeleted nodes (as @shyamvalsan already described, after 30 days)
3. Do not SoftDelete nodes until they will reach an `offline` state, regardless of the 'Ephemeral' flag (no changes required)
4. In Phase 1 the Ephemeral flag should be respected just in the FE. For example: display ephemeral nodes with `Live` or `Stale` state and filter out ephemeral nodes with `offline` state, however, add the ability for users to see those nodes on user action. In this case, dead ephemeral nodes (without data) will not annoy users. In the nearest future, we would be able to expose Alert information for offline nodes (we already have this information on the cloud available, but we are not exposing it right now). This allows people to understand what happened with the node before disconnection.
5. Ephemerality - no effect on ML, Alert, etc. Just an extended filtering opportunity
6. Later on, we might allow users to disable reachability notification for ephemeral nodes only (not for now)

**Comment by cakrit** [comment]
I'm having some trouble following the discussion and would rather discuss in a call. My initial feedback is to keep the following in mind :
- The connection status itself is probably not a good enough criterion to make decisions. Even more important is whether metrics about the node are available to query anywhere (e.g. on a parent). Of course this requires that we finally implement the ability to go back and see charts that are no longer live, potentially from nodes that are no longer live (agent and cloud implementation). We had just two states, queriable and live, but I see we're introducing new states, without clarity IMO on how they map to the queriable (true/false) and live (true/false) states. 
- If we want to properly support ephemeral nodes, or even better, "cattle" views instead of our usual "pet" views, then we need to reconsider more things on the UI. e.g. both the Nodes view and the single node view may not be appropriate at all. I don't have specific ideas about this, I'm just saying that if something starts sounding like too much trouble for the views we have, we can take a step back and ask if it's because the view itself doesn't make sense for this use case.
